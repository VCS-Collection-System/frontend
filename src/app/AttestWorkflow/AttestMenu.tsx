import React, {useEffect, useState} from 'react';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  Stack,
  Text,
} from "@patternfly/react-core";
import { CheckCircleIcon, CogIcon, FileUploadIcon } from '@patternfly/react-icons';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import Feature, { showFeature } from '@app/Components/Feature';
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';

const AttestMenu: React.FunctionComponent = () => {
    const history = useHistory();
    const { keycloak } = useKeycloak();
    const employee = getEmployeeInfo(keycloak);
    const [ isRecovery, setIsRecovery ] = useState(false); 

    useEffect(() => {
      showFeature("recovery", keycloak).then(config => {
        setIsRecovery(config);
      });
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    
    return (
      <Card isRounded>
        <CardWithTitle title="Menu" info="Please select from the following:" />
        <CardBody>
          <Stack hasGutter>
            <Button
              icon={<FileUploadIcon />}
              id="submit-vax-card-menu-button"
              isSmall
              onClick={() => history.push("/attest/vax")}
              variant="tertiary"
            >
              <Text>Submit Vaccination Record</Text>
              <Feature name="proofOptional" renderOff={() => (
                <Text>Supporting Document Upload Required</Text>
              )}>
                &nbsp;
              </Feature>
            </Button>
            <Feature name="covidTest" renderOff={() =>
              <Feature name="recovery"> 
                <Button
                  icon={<FileUploadIcon />}
                  id="submit-vax-test-menu-button"
                  isSmall
                  onClick={() => history.push("/attest/test")}
                  variant="tertiary"
                >
                  <Text>Submit Recovery Certificate</Text>
                  <Text>Supporting Document Upload Required</Text>
                </Button>
              </Feature>
            }> 
              <Button
                icon={<FileUploadIcon />}
                id="submit-vax-test-menu-button"
                isSmall
                onClick={() => history.push("/attest/test")}
                variant="tertiary"
              >
                <Text>Submit Test Results{isRecovery ? " or Recovery Certificate" : ""}</Text>
                <Text>Supporting Document Upload Required</Text>
              </Button>
            </Feature>
            {employee.isApprover ? (
              <Feature name="hrDashboard"> 
                <Button
                  icon={<CheckCircleIcon />}
                  id="validate-submissions-button"
                  isSmall
                  onClick={() => history.push("/hrdashboard")}
                  variant="tertiary"
                >
                  <Text>HR Dashboard</Text>
                  <Text>Review or Update Red Hat VCS Records</Text>
                </Button>
              </Feature>
            ) : ""}
            {employee.isSuperApprover ? (
              <Button
                  icon={<CogIcon />}
                  id="settings-button"
                  isSmall
                  onClick={() => history.push("/settings")}
                  variant="tertiary"
                >
                  <Text>Application Settings</Text>
                  <Text>Update Red Hat VCS Country Settings</Text>
                </Button>
            ) : ""}
          </Stack>
        </CardBody>
        <CardFooter>
          <Button 
            variant="plain" 
            id="attest-menu-back-button"
            onClick={() => history.push("/profile")}
          >
            Back
          </Button>
        </CardFooter>
      </Card>
    );
};
export { AttestMenu }
