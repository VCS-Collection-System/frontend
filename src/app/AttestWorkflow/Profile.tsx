import React from "react";
import {
  Alert,
  Button,
  Card,
  CardBody,
  Checkbox,
  Form,
  FormGroup,
  TextInput,
} from "@patternfly/react-core";
import { CheckCircleIcon } from "@patternfly/react-icons";
import CardWithTitle from "@app/Components/CardWithTitle";
import { useForm } from "react-hook-form";
import { useHistory } from "react-router-dom";
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import { BLOCKED_COUNTRIES } from "@app/utils/constants";

const Profile: React.FunctionComponent = () => {
  const history = useHistory();
  const { handleSubmit } = useForm();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;

  const onSubmit = () => {
    history.push("/attestmenu");
  };

  return (
    <Card isRounded>
      {typeof(employeeCountry) === "undefined" || BLOCKED_COUNTRIES.includes(employeeCountry) ? (
        <Alert variant="danger" title="This application is not currently available in your region." style={{marginBottom: "var(--pf-global--gutter--md)"}}>
          Please visit our <a href="/contact">Help</a> page for more information.
        </Alert>
      ) : ""}
      <CardWithTitle
        title="Profile"
        info="The information below is retrieved from your Red Hat user account."
      />
      <CardBody>
        <Form onSubmit={handleSubmit(onSubmit)}>
          <FormGroup
            id="employee-full-name"
            label="Full Name"
            fieldId="profile-form-employee-full-name"
          >
            <TextInput
              aria-label="employee-full-name"
              isDisabled
              type="text"
              value={(employeeFirstName || "") + " " + (employeeLastName || "")}
            />
          </FormGroup>
          <FormGroup
            id="employee-email-display"
            label="Email"
            fieldId="profile-form-email"
          >
            <TextInput
              aria-label="employee-email-display"
              isDisabled
              type="email"
              value={employeeEmail || ""}
            />
          </FormGroup>
          {employee.isApprover && 
            (
              <Checkbox
                id="reviewer-checkbox"
                label="Red Hat VCS Reviewer"
                aria-label="disabled checked checkbox"
                defaultChecked
                isDisabled
              />
          )}
          <FormGroup fieldId="accept-test-buttons">
            <Button 
              icon={<CheckCircleIcon />} 
              variant="primary" 
              type="submit"
              isDisabled={typeof(employeeCountry) === "undefined" || BLOCKED_COUNTRIES.includes(employeeCountry)}
              >
              Accept
            </Button>
          </FormGroup>
        </Form>
      </CardBody>
    </Card>
  );
};

export { Profile };
