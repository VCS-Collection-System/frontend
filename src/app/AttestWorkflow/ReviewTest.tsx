import React, { useState } from 'react';
import {
  Button, 
  Card,
  CardFooter,
  CardBody,
  Form,
  FormGroup,
  TextInput
} from '@patternfly/react-core';
import { ExternalLinkAltIcon, OutlinedPaperPlaneIcon } from '@patternfly/react-icons';
import { useHistory } from 'react-router-dom';
import CardWithTitle from '@app/Components/CardWithTitle';
import { convertBase64ToBlob, dateConvert } from "@app/utils/utils";
import authaxios from "@app/utils/axiosInterceptor";
import { ResponseErrorModal } from '@app/Components/FeedbackModals';
import Feature from "@app/Components/Feature";
import { useKeycloak } from "@react-keycloak/web";
import { getEmployeeInfo } from '@app/utils/getEmployeeInfo';
import { SUBMIT_TEST_URL } from '@app/utils/constants';

const ReviewTest: React.FunctionComponent = () => {
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false); 
  const [errorState, setErrorState] = useState('');
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeID = employee.id;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;
  const employeeName = (employee.firstName || "") + " " + (employee.lastName || ""); 

  const testImage = sessionStorage.getItem("test_image_file") || "";
  const testImageFilename = sessionStorage.getItem("test_image_filename") || "";
  const altFirstName = sessionStorage.getItem("alt_first_name") || "";
  const altLastName = sessionStorage.getItem("alt_last_name") || "";
  const employeeDOB = sessionStorage.getItem("date_of_birth") || "";
  const covidTestResult = sessionStorage.getItem("test_results") || "";
  const covidTestDate = sessionStorage.getItem("test_date") || "";
  const altName = (altFirstName ? altFirstName : employeeFirstName) + " " +  (altLastName ? altLastName : employeeLastName);

  const handleSubmit = ()=> {
    setIsFormSubmitting(true);
    const testCard = convertBase64ToBlob(testImage)
    const formData = new FormData();

    const employeeInfo = {
      id: employeeID,
      firstName: employeeFirstName,
      lastName: employeeLastName,
      altFirstName: altFirstName,
      altLastName: altLastName,
      dateOfBirth: employeeDOB,
      email: employeeEmail,
      agencyCode: employeeCountry,
      agencyName: employeeCountry,
      divisionCode: ""
    };

    const testInfo = {
      covidTestDate: covidTestDate,
      covidTestResult: covidTestResult.toUpperCase()
    }

    const attachmentExtension = testImageFilename.split('.').pop()?.toLowerCase();
    let attachmentType;

    if(attachmentExtension === "pdf") {
      attachmentType = "application/pdf";
    }
    else if(attachmentExtension === "jpg") {
      attachmentType = "image/jpeg";
    }
    else {
      attachmentType = "image/" + attachmentExtension;
    }

    formData.append('employee', new Blob([JSON.stringify(employeeInfo)], {type: "application/json"}) );
    formData.append('documents', new Blob([JSON.stringify(testInfo)], {type: "application/json"}));
    formData.append("attachment", new Blob([testCard], { type: attachmentType }), "blob."+attachmentExtension);

    authaxios.post(SUBMIT_TEST_URL, formData, {
      headers: {
        'Content-Type': 'multipart/form-data;'
      }
    })
    .then(() => {
      setIsFormSubmitting(false);
      history.push('/thankyou');
    })
    .catch(res => {
      if(res.response) {
        setIsFormSubmitting(false);
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    })
  }

  return (
    <Card isRounded>
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <CardWithTitle
        title="Review Submission"
        info="Review your test result submission and validate that test result image matches displayed data. Please press Submit below to proceed."
      />
      <CardBody>
        <Form>
          <FormGroup label="Person ID" fieldId="review-test-form-employee-id">
            <TextInput
              aria-label="review-test-form-employee-id"
              isDisabled
              type="text"
              value={employeeID}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-test-form-employee-full-name"
            label="Full Name"
          >
            <TextInput
              aria-label="review-test-form-employee-full-name"
              isDisabled
              type="text"
              value={employeeName}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-test-form-employee-alt-name"
            label="Full Preferred Name (as displayed on test result)"
          >
            <TextInput
              aria-label="review-test-form-employee-alt-name"
              isDisabled
              type="text"
              value={altName}
            />
          </FormGroup>
          <Feature name="dateOfBirth">
            <FormGroup
              label="Date of Birth"
              fieldId="review-test-date-of-birth"
            >
              <TextInput
                aria-label="review-test-date-of-birth"
                isDisabled
                type="text"
                value={dateConvert(employeeDOB)}
              />
            </FormGroup>
          </Feature>
          <FormGroup label="Test Results" fieldId="review-test-form-results">
            <TextInput
              aria-label="review-test-form-results"
              isDisabled
              type="text"
              value={sessionStorage.getItem("test_results") || ""}
            />
          </FormGroup>
          <FormGroup label="Test Date" fieldId="review-test-form-date">
            <TextInput
              aria-label="review-test-form-date"
              isDisabled
              type="text"
              value={dateConvert(sessionStorage.getItem("test_date") || "")}
            />
          </FormGroup>
          <FormGroup label="Image preview" fieldId="review-vax-image-preview">
            {testImage && testImageFilename.split('.').pop()?.toLowerCase() == "pdf" &&
              (typeof testImage !== "undefined" ? (
                <Button
                  icon={<ExternalLinkAltIcon />}
                  variant="link"
                  component="a"
                  href={URL.createObjectURL(convertBase64ToBlob(testImage))}
                  target="_blank"
                >
                  Test Result PDF
                </Button>
              ) : null)}
            {testImage && testImageFilename.split('.').pop()?.toLowerCase() != "pdf" &&
              (typeof testImage !== "undefined" ? (
                <div>
                  <img
                    src={URL.createObjectURL(convertBase64ToBlob(testImage))}
                    alt="test-results-image"
                  />
                </div>
              ) : null)}
          </FormGroup>
        </Form>
      </CardBody>
      <CardFooter>
        <Button
          icon={<OutlinedPaperPlaneIcon />}
          isDisabled={isFormSubmitting}
          isLoading={isFormSubmitting}
          onClick={handleSubmit}
          variant="primary"
        >
          Submit
        </Button>
        <Button onClick={() => history.push("/attest/test")} variant="plain">
          Back
        </Button>
      </CardFooter>
    </Card>
  );
};

export { ReviewTest };
