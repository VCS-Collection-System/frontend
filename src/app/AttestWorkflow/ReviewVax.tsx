import React, { useState } from "react";
import {
  Card,
  CardFooter,
  CardBody,
  Button,
  Form,
  FormGroup,
  Grid,
  TextInput
} from "@patternfly/react-core";
import { ExternalLinkAltIcon, OutlinedPaperPlaneIcon } from '@patternfly/react-icons';
import { useHistory } from "react-router-dom";
import { convertBase64ToBlob, dateConvert } from "@app/utils/utils";
import authaxios from "@app/utils/axiosInterceptor";
import { ResponseErrorModal } from '@app/Components/FeedbackModals';
import { useKeycloak } from "@react-keycloak/web";
import CardWithTitle from "@app/Components/CardWithTitle";
import ReviewVaxDose from "@app/Components/ReviewVaxDose";
import Feature from "@app/Components/Feature";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import { SUBMIT_VAX_URL } from "@app/utils/constants"; 

const ReviewVax: React.FunctionComponent = () => {
  const [isErrorModalOpen, setIsErrorModalOpen] = useState(false);
  const [errorState, setErrorState] = useState("");
  const [isFormSubmitting, setIsFormSubmitting] = useState(false);
  const altFirstName = sessionStorage.getItem("alt_first_name") || "";
  const altLastName = sessionStorage.getItem("alt_last_name") || "";
  const employeeDOB = sessionStorage.getItem("date_of_birth") || "";
  const proofType = sessionStorage.getItem("proof_type") || "";
  const vaxImage = sessionStorage.getItem("vax_image") || "";
  const vaxFilename = sessionStorage.getItem("vax_image_filename") || "";

  const doses: {number: number, date: string, lotNumber: string, vaccine: string}[] = [];

  for(let i = 1; sessionStorage.getItem("vax_date_" + i) != null && sessionStorage.getItem("vax_date_" + i) != ""; i++) {
    doses.push({
      number: i,
      date: sessionStorage.getItem("vax_date_" + i) || "",
      lotNumber: sessionStorage.getItem("lot_number_" + i) || "",
      vaccine: sessionStorage.getItem("vaccine_" + i) || ""
    });
  }

  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeID = employee.id;
  const employeeEmail = employee.email;
  const employeeCountry = employee.country;

  const altName = (altFirstName ? altFirstName : employeeFirstName) + " " +  (altLastName ? altLastName : employeeLastName);

  const formSubmitSuccess = () => {
    setIsFormSubmitting(false);
    history.push("/thankyou");
  }
  const handleSubmit = () => {
    setIsFormSubmitting(true);

    const employeeInfo = {
      id: employeeID,
      firstName: employeeFirstName,
      lastName: employeeLastName,
      altFirstName: altFirstName,
      altLastName: altLastName,
      dateOfBirth: employeeDOB,
      email: employeeEmail,
      agencyCode: employeeCountry,
      agencyName: employeeCountry,
      divisionCode: ""
    };

    const vaxInfos = doses.map((dose) => {
      return {
        vaccineBrand: dose.vaccine.toUpperCase(),
        vaccineAdministrationDate: dose.date,
        vaccineShotNumber: dose.number,
        lotNumber: dose.lotNumber,
        proofType: proofType
      };
    });
    
    const formData = new FormData();

    formData.append("employee", new Blob([JSON.stringify(employeeInfo)], { type: "application/json" }));
    formData.append("documents", new Blob([JSON.stringify(vaxInfos)], { type: "application/json" }));
    if(vaxFilename !== "") {
      const vaxCard = convertBase64ToBlob(vaxImage);
      const attachmentExtension = vaxFilename.split('.').pop()?.toLowerCase();
      let attachmentType;

      if(attachmentExtension === "pdf") {
        attachmentType = "application/pdf";
      }
      else if(attachmentExtension === "jpg") {
        attachmentType = "image/jpeg";
      }
      else {
        attachmentType = "image/" + attachmentExtension;
      }

      formData.append("attachment", new Blob([vaxCard], { type: attachmentType }), "blob."+attachmentExtension);
    } else {
      formData.append('attachment', new Blob([], { type: "image/png" }), "blob.png");
    }

    const headers = {
      "Content-Type": "multipart/form-data;",
    };

    authaxios.post(SUBMIT_VAX_URL, formData, {
      headers: headers
    }).then(() => {
      formSubmitSuccess();
    }).catch((res) => {
      if (res.response) {
        setIsFormSubmitting(false);
        setErrorState(res.response.status);
        setIsErrorModalOpen(true);
      }
    });
  }

  const recordOptions = [
    { value: "CDC", label: "CDC Vaccination Card" },
    { value: "DIVOC", label: "Digital Certification of Vaccination" },
    { value: "GREEN_PASS", label: "Green Pass" },
    { value: "OTHER", label: "Other" }
  ];

  const proofPreview = recordOptions.find(option => option.value === proofType) || {};

  return (
    <Card isRounded>
      <ResponseErrorModal 
        isModalOpen={isErrorModalOpen} 
        setIsModalOpen={setIsErrorModalOpen} 
        errorState={errorState}
      />
      <CardWithTitle
        title="Review Vaccination Submission"
        info="Please review your vaccination record submission and validate that card image matches displayed data or it will be rejected. Please press Submit below to proceed."
      />
      <CardBody>
        <Form>
          <FormGroup
            fieldId="review-vax-form-employee-full-name"
            label="Full Name"
          >
            <TextInput
              aria-label="review-vax-form-employee-full-name"
              isDisabled
              type="text"
              value={employeeFirstName + " " + employeeLastName}
            />
          </FormGroup>
          <FormGroup
            fieldId="review-vax-form-employee-alt-name"
            label="Full Preferred Name (as displayed on vaccination record)"
          >
            <TextInput
              aria-label="review-vax-form-employee-alt-name"
              isDisabled
              type="text"
              value={altName}
            />
          </FormGroup>
          <Feature name="dateOfBirth">
            <FormGroup
              label="Date of Birth"
              fieldId="review-vax-form-date-of-birth"
            >
              <TextInput
                aria-label="review-vax-form-date-of-birth"
                isDisabled
                type="text"
                value={dateConvert(employeeDOB)}
              />
            </FormGroup>
          </Feature>
          <Grid hasGutter>
            {doses.map((dose, i) => (
              <ReviewVaxDose doseNumber={dose.number} key={i} date={dateConvert(dose.date)} lotNumber={dose.lotNumber} vaccine={dose.vaccine} />
            ))}
          </Grid>
          <Feature name="proof">
            <FormGroup label="Vaccination Record Type" fieldId="review-vax-form-vaccine-proof">
              <TextInput
                aria-label="review-vax-form-vaccine-proof"
                isDisabled
                type="text"
                value={proofPreview['label']}
              />
            </FormGroup>
            <FormGroup label="Attachment preview" fieldId="review-vax-image-preview">
              {vaxImage && vaxFilename.split('.').pop()?.toLowerCase() == "pdf" && (typeof vaxImage !== "undefined" ? (
                  <Button
                    icon={<ExternalLinkAltIcon />}
                    variant="link"
                    component="a"
                    href={URL.createObjectURL(convertBase64ToBlob(vaxImage))}
                    target="_blank"
                  >
                    Vaccination Card PDF
                  </Button>
              ) : null)}
              {vaxImage && vaxFilename.split('.').pop()?.toLowerCase() != "pdf" &&
                (typeof vaxImage !== "undefined" ? (
                  <div>
                    <img
                      src={URL.createObjectURL(convertBase64ToBlob(vaxImage))}
                      alt="vaccination-card"
                    />
                  </div>
                ) : null)}
              {!vaxImage && (
                <p>No image attached.</p>
              )}
            </FormGroup>
          </Feature>
        </Form>
      </CardBody>
      <CardFooter>
        <Button
          icon={<OutlinedPaperPlaneIcon />}
          isDisabled={isFormSubmitting}
          isLoading={isFormSubmitting}
          variant="primary"
          onClick={handleSubmit}
        >
          Submit
        </Button>
        <Button variant="plain" onClick={() => history.push("/attest/vax")}>
          Back
        </Button>
      </CardFooter>
    </Card>
  );
};

export { ReviewVax };
