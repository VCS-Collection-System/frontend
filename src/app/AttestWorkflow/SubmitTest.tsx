import React, {useEffect, useState} from "react";
import {
  Card,
  CardBody
} from "@patternfly/react-core";
import { useHistory } from "react-router-dom";
import { convertBase64ToBlob, convertBlobToBase64, dateConvert } from "@app/utils/utils";
import CardWithTitle from "@app/Components/CardWithTitle";
import { useKeycloak } from "@react-keycloak/web";
import CovidTestForm from "@app/Components/CovidTestForm";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";
import { showFeature } from "@app/Components/Feature";

const SubmitTest: React.FunctionComponent = () => {
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeID = employee.id;
  const employeeName = (employee.firstName || "") + " " + (employee.lastName || ""); 
  const [ isCovidTest, setIsCovidTest ] = useState(false); 
  const [ isRecovery, setIsRecovery ] = useState(false);

  const onLoad = () => {
    const data = sessionStorage.getItem("test_image_file") || "";

    if (data !== "") {
      return new File([convertBase64ToBlob(data)], sessionStorage.getItem("test_image_filename") || "");
    }

    return "";
  }

  const onSubmit = data => {
    sessionStorage.setItem("alt_first_name", data["alt_first_name"]);
    sessionStorage.setItem("alt_last_name", data["alt_last_name"]);
    sessionStorage.setItem("date_of_birth", dateConvert(data["date_of_birth"]));
    sessionStorage.setItem("test_results", data["test_results"]);
    sessionStorage.setItem("test_date", dateConvert(data["test_date"]));
    sessionStorage.setItem("test_image_filename", data["test_image_file"].name)
    convertBlobToBase64(data["test_image_file"]).then(base64 => {
      sessionStorage["test_image_file"] = base64;
    }).then(() => history.push("/attest/review/test"));
  };

  let cardTitle = "";
  if(isCovidTest) {
    cardTitle += "COVID Test Results";
    if(isRecovery) {
      cardTitle += " or Certificate of Recovery";
    }
  } else {
    cardTitle += "Certificate of Recovery";
  }

  useEffect(() => {
    showFeature("covidTest", keycloak).then(config => {
      setIsCovidTest(config);
    });

    showFeature("recovery", keycloak).then(config => {
      setIsRecovery(config);
    });
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Card isRounded>
      <CardWithTitle title={cardTitle} info="Submit Supporting Documentation" />
      <CardBody>
        <CovidTestForm
          isHRSubmit={false}
          employeeId={employeeID}
          employeeName={employeeName}
          onSubmit={onSubmit}
          onLoad={onLoad}
        />
      </CardBody>
    </Card>
  );
};

export { SubmitTest };