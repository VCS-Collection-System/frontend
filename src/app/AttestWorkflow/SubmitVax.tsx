import React from "react";
import {
  Card,
  CardBody
} from "@patternfly/react-core";
import { useHistory } from "react-router-dom";
import { convertBase64ToBlob, convertBlobToBase64, dateConvert } from "@app/utils/utils";
import { useKeycloak } from "@react-keycloak/web";
import CardWithTitle from "@app/Components/CardWithTitle";
import CovidVaxForm from "@app/Components/CovidVaxForm";
import { getEmployeeInfo } from "@app/utils/getEmployeeInfo";

const SubmitVax: React.FunctionComponent = () => {
  const history = useHistory();
  const { keycloak } = useKeycloak();
  const employee = getEmployeeInfo(keycloak);
  const employeeFirstName = employee.firstName;
  const employeeLastName = employee.lastName;
  const employeeID = employee.id;

  const onLoad = () => {
    const data = sessionStorage.getItem("vax_image") || "";

    if (data !== "") {
      return new File([convertBase64ToBlob(data)], sessionStorage.getItem("vax_image_filename") || "");
    }

    return "";
  }

  const onSubmit = (data) => {
    sessionStorage.setItem("alt_first_name", data["alt_first_name"]);
    sessionStorage.setItem("alt_last_name", data["alt_last_name"]);
    sessionStorage.setItem("date_of_birth", dateConvert(data["date_of_birth"]));
    sessionStorage.setItem("vaccine", data["vaccine"]);
    sessionStorage.setItem("proof_type", data["proof_type"] || "OTHER");
    sessionStorage.setItem("vax_image_filename", data["vax_image"]?.name || "");
    for(let i = 1; typeof(data["vax_date_" + i]) !== "undefined"; i++) {
      sessionStorage.setItem("vax_date_" + i, dateConvert(data["vax_date_" + i]));
      sessionStorage.setItem("lot_number_" + i, data["lot_number_" + i]);
      sessionStorage.setItem("vaccine_" + i, data["vaccine_" + i]);
    }
    if(data["vax_image"]?.name) {
      convertBlobToBase64(data["vax_image"]).then((base64) => {
        sessionStorage["vax_image"] = base64;
      }).then(()=> history.push("/attest/review/vax"));
    } else {
      history.push("/attest/review/vax");
    }
  };

  return (
    <Card isRounded>
      <CardWithTitle title="Vaccination Form" info="Enter vaccination info" />
      <CardBody>
        <CovidVaxForm 
          isHRSubmit={false}
          employeeId={employeeID || ""}
          employeeName={(employeeFirstName || "") + " " + (employeeLastName || "")}
          onSubmit={onSubmit}
          onLoad={onLoad}
        />
      </CardBody>
    </Card>
  );
};

export { SubmitVax };
