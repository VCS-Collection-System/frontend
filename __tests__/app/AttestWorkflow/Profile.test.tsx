import React from "react"
import { Profile } from "../../../src/app/AttestWorkflow/Profile"
import { act, fireEvent, render, screen } from '@testing-library/react'
import mockKeycloak from "../../../__mocks__/keycloakMock"
import reactRouterDom from "react-router-dom"

jest.mock("@react-keycloak/web", () => ({
  useKeycloak: jest.fn(() => ({ keycloak: mockKeycloak })),
}))
jest.mock("@app/utils/constants", () => ({
  KEYCLOAK_FIRST_NAME: "firstName",
  KEYCLOAK_LAST_NAME: "lastName",
  KEYCLOAK_EMAIL: "email",
  KEYCLOAK_COUNTRY: "attributes.agency[0]",
  KEYCLOAK_ID: "attributes.workforceid[0]",
  APPROVER_ROLES: ["mock-approver"],
  BLOCKED_COUNTRIES: []
}))
jest.mock("react-router-dom")

const mockPush = jest.fn()
reactRouterDom.useHistory = jest.fn().mockReturnValue({ push: mockPush })

describe('<Profile />', ()=> {
  it('should be defined', () => {
      expect(Profile).toBeDefined()
  })
  it("should match existing snapshot", () => {
      expect(render(<Profile />).asFragment()).toMatchSnapshot()
  })
  it("should be able to go attestmenu", async () => {
    mockKeycloak.profile = {
      firstName: "mockFirst",
      lastName: "mockLast",
      email: "mock@email.com",
      attributes: {
          workforceid: ["mockId"],
          agency: ["USA"]
      }
    }
    await act(async () => {
      render(<Profile />)
    })
    const acceptButton = await screen.getByRole('button', {name: /Accept/i})
    await fireEvent.click(acceptButton)
    expect(mockPush).toHaveBeenCalledWith('/attestmenu')
  })

  it("Empty values should not return undefined", async () => {
    mockKeycloak.profile = {
      firstName: "",
      lastName: "",
      email: "",
      attributes: {
          workforceid: ["mockId"],
          agency: ["USA"]
      }
    }
    await act(async () => {
      render(<Profile />)
    })
    expect(screen.queryAllByText('undefined')).toEqual([])
    expect(screen.getByRole('textbox', {name: /employee-full-name/i}).getAttribute('value')).toEqual(" ")
    expect(screen.getByRole('textbox', {name: /employee-email-display/i}).getAttribute('value')).toEqual("")
  })

  it("Should render caution banner when not in allowed countries", async () => {
    mockKeycloak.profile = {
      firstName: "",
      lastName: "",
      email: "",
      attributes: {
          workforceid: ["mockId"],
          agency: ["FAKE"]
      }
    }
    await act(async () => {
      render(<Profile />)
    })
    expect(screen.queryAllByText('This application is not currently available in your region.')).toBeTruthy()
  })

  it("should display approver role when is an approver", async () => {
    mockKeycloak.hasRealmRole.mockReturnValue(true)
    await act(async () => {
      render(<Profile />)
    })
    expect(screen.getByText('Red Hat VCS Reviewer')).toBeTruthy();
  })

  it("should not display approver role when not an approver", async () => {
    mockKeycloak.hasRealmRole.mockReturnValue(false)
    await act(async () => {
      render(<Profile />)
    })
    expect(screen.queryAllByText('Red Hat VCS Reviewer')).toEqual([])
  })
})