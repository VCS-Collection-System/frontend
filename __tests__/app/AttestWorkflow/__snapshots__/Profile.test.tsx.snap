// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> should match existing snapshot 1`] = `
<DocumentFragment>
  <article
    class="pf-c-card pf-m-rounded"
    data-ouia-component-id="OUIA-Generated-Card-1"
    data-ouia-component-type="PF4/Card"
    data-ouia-safe="true"
    id=""
  >
    <div
      class="pf-c-card__title"
    >
      <p
        class=""
        data-pf-content="true"
      >
        Profile
      </p>
      <div
        class="pf-c-content"
      >
        <small
          class=""
          data-pf-content="true"
        >
          The information below is retrieved from your Red Hat user account.
        </small>
      </div>
    </div>
    <div
      class="pf-c-card__body"
    >
      <form
        class="pf-c-form"
        novalidate=""
      >
        <div
          class="pf-c-form__group"
          id="employee-full-name"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="profile-form-employee-full-name"
            >
              <span
                class="pf-c-form__label-text"
              >
                Full Name
              </span>
            </label>
             
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <input
              aria-invalid="false"
              aria-label="employee-full-name"
              class="pf-c-form-control"
              disabled=""
              type="text"
              value="mockFirst mockLast"
            />
          </div>
        </div>
        <div
          class="pf-c-form__group"
          id="employee-email-display"
        >
          <div
            class="pf-c-form__group-label"
          >
            <label
              class="pf-c-form__label"
              for="profile-form-email"
            >
              <span
                class="pf-c-form__label-text"
              >
                Email
              </span>
            </label>
             
          </div>
          <div
            class="pf-c-form__group-control"
          >
            <input
              aria-invalid="false"
              aria-label="employee-email-display"
              class="pf-c-form-control"
              disabled=""
              type="email"
              value="mock@email.com"
            />
          </div>
        </div>
        <div
          class="pf-c-form__group"
        >
          <div
            class="pf-c-form__group-control"
          >
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-primary"
              data-ouia-component-id="OUIA-Generated-Button-primary-1"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="submit"
            >
              <span
                class="pf-c-button__icon pf-m-start"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
                  />
                </svg>
              </span>
              Accept
            </button>
          </div>
        </div>
      </form>
    </div>
  </article>
</DocumentFragment>
`;
